File: sdk/python/inkbox/tunnels/client/_bootstrap.py
"""
inkbox/tunnels/client/_bootstrap.py

Pre-runtime orchestration: state-file lookup, server lookup, optional
CSR + sign for passthrough. Returns a fully-resolved bundle the runtime
can connect with.

Tunnels are provisioned atomically by ``inkbox.create_identity(...)``;
``connect`` is read-only on the control plane (with the exception of
passthrough CSR signing). Data-plane authentication uses the client's
API key — no per-tunnel secret is involved.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from pathlib import Path
from typing import Sequence
from uuid import UUID

from inkbox.exceptions import InkboxAPIError
from inkbox.tunnels._validation import validate_tunnel_name
from inkbox.tunnels.client._cert import (
    build_csr,
    cert_needs_sign,
    key_pem_bytes,
    load_or_create_keypair,
    write_cert_chain,
)
from inkbox.tunnels.client._state import (
    CERT_FILE,
    StateEntry,
    ensure_private_state_dir,
    load_state,
    save_state,
)
from inkbox.tunnels.client._tls import TLSTerminator
from inkbox.tunnels.exceptions import (
    TunnelNotProvisioned,
    TunnelRemoved,
    TunnelStateConflict,
)
from inkbox.tunnels.resources.tunnels import POOL_SIZE_MAX, POOL_SIZE_MIN
from inkbox.tunnels.types import TLSMode, Tunnel, TunnelStatus

logger = logging.getLogger("inkbox.tunnels")


# Default tunnel zone — used as a fallback when neither the server
# response nor the state file specifies one.
PROD_ZONE = "inkboxwire.com"


@dataclass
class TunnelBundle:
    tunnel: Tunnel
    public_host: str
    zone: str
    tls_terminator: TLSTerminator | None


def _status_repr(status: TunnelStatus | str) -> str:
    """Stringify a status that may be either an enum or a raw server string."""
    if isinstance(status, TunnelStatus):
        return status.value
    return str(status)


def validate_pool_size(pool_size: int | None) -> None:
    if pool_size is None:
        return
    if not isinstance(pool_size, int) or pool_size < POOL_SIZE_MIN:
        raise ValueError(
            f"pool_size must be an int >= {POOL_SIZE_MIN} (got {pool_size!r})",
        )
    if pool_size > POOL_SIZE_MAX:
        raise ValueError(
            f"pool_size must be <= {POOL_SIZE_MAX} (got {pool_size!r})",
        )


def resolve_zone_and_host(
    *,
    name: str,
    server_zone: str | None,
    server_public_host: str | None,
    state: StateEntry | None,
    data_plane_zone_override: str | None,
) -> tuple[str, str]:
    """Pick the zone + public host using the documented precedence.

    ``data_plane_zone_override`` only overrides the zone (data-plane h2
    endpoint). ``public_host`` always comes from server > state >
    prod-zone fallback.
    """
    if server_public_host:
        public_host = server_public_host
    elif state and state.public_host:
        public_host = state.public_host
    else:
        public_host = f"{name}.{PROD_ZONE}"

    if data_plane_zone_override:
        zone = data_plane_zone_override
    elif server_zone:
        zone = server_zone
    elif state and state.zone:
        zone = state.zone
    else:
        zone = PROD_ZONE
    return zone, public_host


def bootstrap(
    *,
    inkbox: object,  # Inkbox instance; circular-import-friendly typing
    name: str,
    state_dir: Path,
    data_plane_zone_override: str | None,
    alpn_protocols: Sequence[str] = ("http/1.1",),
) -> TunnelBundle:
    """Resolve a tunnel for ``connect()``: server lookup + cert dance for
    passthrough."""
    validate_tunnel_name(name)

    state_dir = Path(state_dir).expanduser()
    ensure_private_state_dir(state_dir)
    state = load_state(state_dir)

    tunnels = inkbox.tunnels  # type: ignore[attr-defined]

    tunnel: Tunnel | None = None
    state_tunnel_id: UUID | None = None
    if state and state.tunnel_id:
        try:
            state_tunnel_id = UUID(state.tunnel_id)
        except ValueError:
            state_tunnel_id = None

    if state_tunnel_id is not None:
        try:
            tunnel = tunnels.get(state_tunnel_id)
        except InkboxAPIError as err:
            if err.status_code == 404:
                raise TunnelRemoved(
                    f"tunnel {name!r} (id={state_tunnel_id}) has been removed; "
                    f"clear {state_dir} and call inkbox.create_identity({name!r}) "
                    "to start fresh",
                ) from err
            raise

    if tunnel is None:
        # Look up by name (the server lists only live tunnels).
        for t in tunnels.list():
            if t.tunnel_name == name:
                tunnel = t
                break

    if tunnel is None:
        raise TunnelNotProvisioned(
            f"no tunnel named {name!r} exists in this org. "
            "Tunnels are provisioned atomically by "
            f"inkbox.create_identity({name!r}); call that first."
        )

    # Edge tunnels must be ACTIVE before we open the data plane.
    # Passthrough has its own AWAITING_CERT branch below.
    if (
        tunnel.tls_mode == TLSMode.EDGE
        and tunnel.status != TunnelStatus.ACTIVE
    ):
        raise TunnelStateConflict(
            status_code=409,
            detail=(
                f"tunnel {name!r} is in status {_status_repr(tunnel.status)}; "
                "expected active before opening the data plane"
            ),
        )

    # Cert dance for passthrough.
    terminator: TLSTerminator | None = None
    if tunnel.tls_mode == TLSMode.PASSTHROUGH:
        zone, public_host = resolve_zone_and_host(
            name=name,
            server_zone=tunnel.zone,
            server_public_host=tunnel.public_host,
            state=state,
            data_plane_zone_override=data_plane_zone_override,
        )
        key = load_or_create_keypair(state_dir)
        if (
            tunnel.status == TunnelStatus.AWAITING_CERT
            or cert_needs_sign(state_dir, key)
        ):
            csr_pem = build_csr(key, public_host)
            logger.info("POST /tunnels/%s/sign-csr", tunnel.id)
            signed = tunnels.sign_csr(tunnel.id, csr_pem=csr_pem)
            chain_bytes = write_cert_chain(
                state_dir, signed.cert_pem, signed.chain_pem,
            )
            # Refresh tunnel record to pick up the new active status.
            tunnel = tunnels.get(tunnel.id)
        else:
            chain_bytes = (state_dir / CERT_FILE).read_bytes()

        if tunnel.status != TunnelStatus.ACTIVE:
            raise TunnelStateConflict(
                status_code=409,
                detail=(
                    f"tunnel {name!r} is in status {_status_repr(tunnel.status)}; "
                    "expected active after CSR sign"
                ),
            )

        terminator = TLSTerminator(
            cert_chain_pem=chain_bytes,
            key_pem=key_pem_bytes(key),
            alpn_protocols=alpn_protocols,
        )

    zone, public_host = resolve_zone_and_host(
        name=name,
        server_zone=tunnel.zone,
        server_public_host=tunnel.public_host,
        state=state,
        data_plane_zone_override=data_plane_zone_override,
    )

    # Persist final state (zone/public_host learned from server).
    save_state(state_dir, StateEntry(
        tunnel_id=str(tunnel.id),
        name=name,
        mode=tunnel.tls_mode.value,
        zone=zone,
        public_host=public_host,
    ))

    return TunnelBundle(
        tunnel=tunnel,
        public_host=public_host,
        zone=zone,
        tls_terminator=terminator,
    )

File: sdk/rust/src/tunnels/client/bootstrap.rs
//! Pre-runtime orchestration: state-file lookup, server lookup, optional CSR
//! sign for passthrough. Resolves a [`TunnelBundle`] the runtime connects with.
//!
//! Tunnels are provisioned atomically by `create_identity(...)`; `connect` is
//! read-only on the control plane (except passthrough CSR signing). Data-plane
//! auth uses the client's API key — no per-tunnel secret is involved.
//!
//! Ported from `inkbox/tunnels/client/_bootstrap.py`.

use crate::error::{InkboxError, Result};
use crate::tunnels::resources::{POOL_SIZE_MAX, POOL_SIZE_MIN};
use crate::tunnels::types::Tunnel;

use super::state::StateEntry;

/// Default tunnel zone — fallback when neither the server response nor the
/// state file specifies one.
pub const PROD_ZONE: &str = "inkboxwire.com";

/// Fully-resolved tunnel ready to connect. The TLS terminator (passthrough)
/// is represented by its cert-chain + key PEM bytes; the runtime builds the
/// rustls config from them.
#[derive(Debug, Clone)]
pub struct TunnelBundle {
    pub tunnel: Tunnel,
    pub public_host: String,
    pub zone: String,
    /// Passthrough only: `(cert_chain_pem, key_pem)`. `None` for edge tunnels.
    pub tls_material: Option<(Vec<u8>, Vec<u8>)>,
}

/// Validate the optional `pool_size` kwarg (1..=32). Mirrors Python's
/// `validate_pool_size`, raising on out-of-range values.
pub fn validate_pool_size(pool_size: Option<i64>) -> Result<()> {
    match pool_size {
        None => Ok(()),
        Some(n) if n < POOL_SIZE_MIN => Err(InkboxError::InvalidArgument(format!(
            "pool_size must be an int >= {POOL_SIZE_MIN} (got {n})"
        ))),
        Some(n) if n > POOL_SIZE_MAX => Err(InkboxError::InvalidArgument(format!(
            "pool_size must be <= {POOL_SIZE_MAX} (got {n})"
        ))),
        Some(_) => Ok(()),
    }
}

/// Pick the zone + public host using the documented precedence.
///
/// `data_plane_zone_override` only overrides the zone (the data-plane h2
/// endpoint). `public_host` always comes from server > state > prod-zone
/// fallback.
///
/// # Arguments
/// * `name` - The tunnel name (= owning identity's agent handle).
/// * `server_zone` / `server_public_host` - Values from the control-plane
///   tunnel record (may be empty).
/// * `state` - The on-disk state entry, if any.
/// * `data_plane_zone_override` - Expert-only zone override.
///
/// # Returns
/// `(zone, public_host)`.
pub fn resolve_zone_and_host(
    name: &str,
    server_zone: Option<&str>,
    server_public_host: Option<&str>,
    state: Option<&StateEntry>,
    data_plane_zone_override: Option<&str>,
) -> (String, String) {
    let public_host = if let Some(h) = server_public_host.filter(|s| !s.is_empty()) {
        h.to_string()
    } else if let Some(h) = state
        .and_then(|s| s.public_host.as_deref())
        .filter(|s| !s.is_empty())
    {
        h.to_string()
    } else {
        format!("{name}.{PROD_ZONE}")
    };

    let zone = if let Some(z) = data_plane_zone_override.filter(|s| !s.is_empty()) {
        z.to_string()
    } else if let Some(z) = server_zone.filter(|s| !s.is_empty()) {
        z.to_string()
    } else if let Some(z) = state
        .and_then(|s| s.zone.as_deref())
        .filter(|s| !s.is_empty())
    {
        z.to_string()
    } else {
        PROD_ZONE.to_string()
    };
    (zone, public_host)
}

// The full `bootstrap` flow (state lookup → server get/list → passthrough
// CSR dance → resolve_zone_and_host → save_state → TunnelBundle) is driven
// from `crate::tunnels::resources::tunnels::TunnelsResource::connect`, which
// holds the sync control-plane transport this module lacks. The pure,
// testable pieces (`validate_pool_size`, `resolve_zone_and_host`) live here.

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pool_size_validation() {
        assert!(validate_pool_size(None).is_ok());
        assert!(validate_pool_size(Some(1)).is_ok());
        assert!(validate_pool_size(Some(32)).is_ok());
        assert!(validate_pool_size(Some(0)).is_err());
        assert!(validate_pool_size(Some(-1)).is_err());
        assert!(validate_pool_size(Some(33)).is_err());
    }

    #[test]
    fn zone_host_server_wins() {
        let (zone, host) = resolve_zone_and_host(
            "my-agent",
            Some("zone.example"),
            Some("my-agent.zone.example"),
            None,
            None,
        );
        assert_eq!(zone, "zone.example");
        assert_eq!(host, "my-agent.zone.example");
    }

    #[test]
    fn zone_host_fallback_to_prod() {
        let (zone, host) = resolve_zone_and_host("my-agent", None, None, None, None);
        assert_eq!(zone, PROD_ZONE);
        assert_eq!(host, format!("my-agent.{PROD_ZONE}"));
    }

    #[test]
    fn zone_override_only_changes_zone() {
        let (zone, host) = resolve_zone_and_host(
            "my-agent",
            Some("server.example"),
            Some("my-agent.server.example"),
            None,
            Some("override.example"),
        );
        assert_eq!(zone, "override.example");
        // public_host still comes from server.
        assert_eq!(host, "my-agent.server.example");
    }

    #[test]
    fn state_used_when_server_empty() {
        let state = StateEntry {
            tunnel_id: "id".into(),
            name: "my-agent".into(),
            mode: Some("edge".into()),
            zone: Some("state.example".into()),
            public_host: Some("my-agent.state.example".into()),
        };
        let (zone, host) = resolve_zone_and_host("my-agent", None, None, Some(&state), None);
        assert_eq!(zone, "state.example");
        assert_eq!(host, "my-agent.state.example");
    }
}
